main.go

```diff
@@ -0,0 +1,252 @@
+	elasticsearchUser = cfg.GetString("elasticsearch.user")
+	elasticsearchPassword = cfg.GetString("elasticsearch.password")
+	elasticsearchIndex = cfg.GetString("elasticsearch.index")
+	max, err := strconv.Atoi(cfg.GetString("infosquito.maximum_in_prefix"))
+	if err != nil {
+		log.Fatal("Couldn't parse integer out of infosquito.maximum_in_prefix")
+	}
+	maxInPrefix = max
+
+	base, err := strconv.Atoi(cfg.GetString("infosquito.base_prefix_length"))
+	if err != nil {
+		log.Fatal("Couldn't parse integer out of infosquito.base_prefix_length")
+	}
+	basePrefixLength = base
+}
+
+func loadAMQPConfig() {
+	amqpURI = cfg.GetString("amqp.uri")
+	amqpExchangeName = cfg.GetString("amqp.exchange.name")
+	amqpExchangeType = cfg.GetString("amqp.exchange.type")
+	amqpQueuePrefix = cfg.GetString("amqp.queue_prefix")
+}
+
+func getQueueName(prefix string) string {
+	if len(prefix) > 0 {
+		return fmt.Sprintf("%s.infosquito2", prefix)
+	}
+	return "infosquito2"
+}
+
+func generatePrefixes(length int) []string {
+	prefixes := int(math.Pow(16, float64(length)))
+	res := make([]string, prefixes)
+	for i := 0; i < prefixes; i++ {
+		res[i] = fmt.Sprintf("%0"+strconv.Itoa(length)+"x", i)
+	}
+	return res
+}
+
+func splitPrefix(prefix string) []string {
+	res := make([]string, 16)
+	for i := 0; i < 16; i++ {
+		res[i] = fmt.Sprintf("%s%x", prefix, i)
+	}
+	return res
+}
+
+func TryReindexPrefix(db *ICATConnection, es *ESConnection, prefix string) error {
+	err := ReindexPrefix(db, es, prefix)
+	if err == ErrTooManyResults {
+		for _, newprefix := range splitPrefix(prefix) {
+			err = TryReindexPrefix(db, es, newprefix)
+			if err != nil {
+				return err
+			}
+		}
+	} else if err != nil {
+		return err
+	}
+	return nil
+}
+
+func main() {
+	checkMode()
+	initConfig(*cfgPath)
+
+	db, err := SetupDB(dbURI)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	es, err := SetupES(elasticsearchBase, elasticsearchUser, elasticsearchPassword, elasticsearchIndex)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if *mode == "full" {
+		log.Info("Full indexing mode selected.")
+		// do full mode
+		for _, prefix := range generatePrefixes(basePrefixLength) {
+			log.Infof("Reindexing prefix %s", prefix)
+			err := TryReindexPrefix(db, es, prefix)
+			if err != nil {
+				log.Fatal(err)
+			}
+		}
+		return
+	}
+
+	// periodic mode
+	log.Info("Periodic indexing mode selected.")
+	loadAMQPConfig()
+
+	listenClient, err := messaging.NewClient(amqpURI, true)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer listenClient.Close()
+
+	publishClient, err := messaging.NewClient(amqpURI, true)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer publishClient.Close()
+
+	publishClient.SetupPublishing(amqpExchangeName)
+
+	go listenClient.Listen()
+
+	queueName := getQueueName(amqpQueuePrefix)
+	listenClient.AddConsumerMulti(
+		amqpExchangeName,
+		amqpExchangeType,
+		queueName,
+		[]string{"index.all", "index.data", "index.data.prefix.#"},
+		func(del amqp.Delivery) {
```

Comment:
Yeah, that's probably a good idea, or maybe I'll keep it anonymous but break out bits of it. I figure since it essentially does two things (generate prefixes and send messages vs. try to reindex and send messages if needed) maybe those should be separate too.

Reply:
That makes sense.